src/util/memstream.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Tokio compatible in-memory stream implementation. Helpful for testing.
+
+// TODO(povilas): move to future_utils crate - this module is generic and highly reusable.
+
+use priv_prelude::*;
+use std::fmt;
+use std::io::{self, Read, Write};
+
+/// In-memory stream implementing `AsyncWrite + AsyncRead`. This stream is not connected with
+/// any other stream. Meaning what you write to the stream you can later read from it.
+#[derive(Default)]
+pub struct EchoStream {
+    buf: Vec<u8>,
+    shutdown: bool,
+}
+
+impl Read for EchoStream {
+    /// Reads as much data as possible to the given buffer and removes that data from internal
+    /// buffer.
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if self.buf.is_empty() {
+            if self.shutdown {
+                return Ok(0);
+            } else {
+                return Err(io::ErrorKind::WouldBlock.into());
+            }
+        }
+
+        let mut bytes_read = 0;
+        while bytes_read < buf.len() && !self.buf.is_empty() {
+            buf[bytes_read] = self.buf.remove(0);
+            bytes_read += 1;
+        }
```

Comment:
Seems a little primitive being done this way and also not optimal - removing elt 1 by 1 from the beginning of the vector is not the most efficient thing to do, esp if you have information already about how many to remove which can make the impl do so optimisations. Something that came to mind 1st without much thought (so if you find something better then go for it) is the following:
```
let min = cmp::min(self.buf.len(), buf.len());
buf.iter_mut().zip(self.buf.drain(..min)).for_each(|(dst, src)| *dst = src);
```
should do it (and probably cleaner, safer and cripier) ?

Guess mem copy is even quicker but probably/unfortunately `unsafe` ? Anyway you can evaluate further.

Reply:
Yep, that was the least complex solution though, didn;t bother to optimize since that's for tests. If you think it's worth it, I'll do it :)

Reply:
Ah the quality of code shouldn't be different in production and in test - i would change it. Test is not a second class citizen when it comes to code quality, sure it can have asserts and other things not used in production OR some magic number (e.g., vec with 4 elements etc) , but not something that does not conform to idiomatic usage of Rust for e.g.